lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/REEFEnvironment.java

```diff
@@ -0,0 +1,154 @@
+
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * The main entry point into any REEF process (Driver and Evaluator).
+ * It is mostly reading from the command line to instantiate
+ * the runtime clock and calling .run() on it.
+ */
+public final class REEFEnvironment implements Runnable, AutoCloseable {
+
+  /**
+   * Parameter to enable Wake network profiling. By default profiling is disabled.
+   */
+  @NamedParameter(doc = "If true, profiling will be enabled", short_name = "profiling", default_value = "false")
+  private static final class ProfilingEnabled implements Name<Boolean> { }
+
+  private static final Logger LOG = Logger.getLogger(REEFEnvironment.class.getName());
+
+  private static final Tang TANG = Tang.Factory.getTang();
+
+  /** Main event loop of current REEF component (Driver or Evaluator). */
+  private final Clock clock;
+
+  /** Error handler that processes all uncaught REEF exceptions. */
+  private final REEFErrorHandler errorHandler;
+
+  /**
+   * Create new REEF environment.
+   * @param config REEF component (Driver or Evaluator) configuration.
+   * It is usually read from config file by REEFLauncher.
+   * @throws InjectionException Thrown on configuration error.
+   */
+  @SuppressWarnings("checkstyle:illegalcatch")
+  public REEFEnvironment(final Configuration config) throws InjectionException {
+
+    if (LOG.isLoggable(Level.FINEST)) {
+      LOG.log(Level.FINEST, "Configuration:\n--\n{0}\n--",
+          new AvroConfigurationSerializer().toString(config, true));
+    }
+
+    final Injector injector = TANG.newInjector(config);
+
+    if (injector.getNamedInstance(ProfilingEnabled.class)) {
+      final WakeProfiler profiler = new WakeProfiler();
+      ProfilingStopHandler.setProfiler(profiler);
+      injector.bindAspect(profiler);
+    }
+
+    injector.getInstance(REEFVersion.class).logVersion();
+
+    this.errorHandler = injector.getInstance(REEFErrorHandler.class);
+
+    try {
+      this.clock = injector.getInstance(Clock.class);
+    } catch (final Throwable ex) {
+      LOG.log(Level.SEVERE, "Error while instantiating the clock", ex);
+      try {
+        this.errorHandler.onNext(ex);
+      } catch (final Throwable exHandling) {
+        LOG.log(Level.SEVERE, "Error while handling the exception " + ex, exHandling);
+      }
+      throw ex;
+    }
+  }
+
+  /**
+   * Close and cleanup the environment.
+   * Invoke .close() on all closeable members (clock and error handler).
+   */
+  @Override
+  @SuppressWarnings("checkstyle:illegalcatch")
+  public void close() {
+
+    LOG.log(Level.FINER, "Closing REEF Environment - start");
+
+    try {
+      this.clock.close();
+    } catch (final Throwable ex) {
+      LOG.log(Level.SEVERE, "Error while closing the clock", ex);
+      try {
+        this.errorHandler.onNext(ex);
+      } catch (final Throwable exHandling) {
+        LOG.log(Level.SEVERE, "Error while handling the exception " + ex, exHandling);
```

Comment:
General question on these log lines: Do we make sure we log the exceptions exactly once here?